// unocss-preset-weapp@0.1.18/dist/rules.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const decoration = require('./shared/unocss-preset-weapp.41ef2805.cjs');
const _default = require('./shared/unocss-preset-weapp.0342cae8.cjs');
require('unplugin-transform-class/utils');
require('./shared/unocss-preset-weapp.80bd545b.cjs');
require('@unocss/core');



exports.alignments = decoration.alignments;
exports.animations = decoration.animations;
exports.appearance = decoration.appearance;
exports.appearances = decoration.appearances;
exports.backgroundBlendModes = decoration.backgroundBlendModes;
exports.backgroundStyles = decoration.backgroundStyles;
exports.borderStyles = decoration.borderStyles;
exports.borders = decoration.borders;
exports.boxShadows = decoration.boxShadows;
exports.boxShadowsBase = decoration.boxShadowsBase;
exports.boxSizing = decoration.boxSizing;
exports.breaks = decoration.breaks;
exports.contentVisibility = decoration.contentVisibility;
exports.contents = decoration.contents;
exports.cursors = decoration.cursors;
exports.displays = decoration.displays;
exports.floats = decoration.floats;
exports.fontSmoothings = decoration.fontSmoothings;
exports.fontStyles = decoration.fontStyles;
exports.fonts = decoration.fonts;
exports.hyphens = decoration.hyphens;
exports.insets = decoration.insets;
exports.isolations = decoration.isolations;
exports.justifies = decoration.justifies;
exports.mixBlendModes = decoration.mixBlendModes;
exports.objectPositions = decoration.objectPositions;
exports.orders = decoration.orders;
exports.outline = decoration.outline;
exports.overflows = decoration.overflows;
exports.placements = decoration.placements;
exports.pointerEvents = decoration.pointerEvents;
exports.positions = decoration.positions;
exports.resizes = decoration.resizes;
exports.ringBase = decoration.ringBase;
exports.rings = decoration.rings;
exports.screenReadersAccess = decoration.screenReadersAccess;
exports.tabSizes = decoration.tabSizes;
exports.textAligns = decoration.textAligns;
exports.textDecorations = decoration.textDecorations;
exports.textIndents = decoration.textIndents;
exports.textOverflows = decoration.textOverflows;
exports.textShadows = decoration.textShadows;
exports.textStrokes = decoration.textStrokes;
exports.textTransforms = decoration.textTransforms;
exports.transformBase = decoration.transformBase;
exports.transforms = decoration.transforms;
exports.transitions = decoration.transitions;
exports.userSelects = decoration.userSelects;
exports.varEmpty = decoration.varEmpty;
exports.verticalAligns = decoration.verticalAligns;
exports.whitespaces = decoration.whitespaces;
exports.willChange = decoration.willChange;
exports.writingModes = decoration.writingModes;
exports.writingOrientations = decoration.writingOrientations;
exports.zIndexes = decoration.zIndexes;
exports.aspectRatio = _default.aspectRatio;
exports.bgColors = _default.bgColors;
exports.cssProperty = _default.cssProperty;
exports.cssVariables = _default.cssVariables;
exports.flex = _default.flex;
exports.gaps = _default.gaps;
exports.grids = _default.grids;
exports.margins = _default.margins;
exports.opacity = _default.opacity;
exports.paddings = _default.paddings;
exports.questionMark = _default.questionMark;
exports.rules = _default.rules;
exports.safeArea = _default.safeArea;
exports.sizes = _default.sizes;
exports.svgUtilities = _default.svgUtilities;
exports.textColors = _default.textColors;
