// unocss-preset-weapp/dist/utils.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const colors = require('./shared/unocss-preset-weapp.80bd545b.cjs');
const variants = require('./shared/unocss-preset-weapp.60967b4e.cjs');
require('@unocss/core');
require('unplugin-transform-class/utils');



exports.CONTROL_MINI_NO_NEGATIVE = colors.CONTROL_MINI_NO_NEGATIVE;
exports.colorOpacityToString = colors.colorOpacityToString;
exports.colorResolver = colors.colorResolver;
exports.colorToString = colors.colorToString;
exports.colorableShadows = colors.colorableShadows;
exports.cornerMap = colors.cornerMap;
exports.directionMap = colors.directionMap;
exports.directionSize = colors.directionSize;
exports.getComponent = colors.getComponent;
exports.getComponents = colors.getComponents;
exports.globalKeywords = colors.globalKeywords;
exports.h = colors.h;
exports.handler = colors.handler;
exports.hasParseableColor = colors.hasParseableColor;
exports.hex2rgba = colors.hex2rgba;
exports.insetMap = colors.insetMap;
exports.makeGlobalStaticRules = colors.makeGlobalStaticRules;
exports.parseColor = colors.parseColor;
exports.parseCssColor = colors.parseCssColor;
exports.positionMap = colors.positionMap;
exports.resolveBreakpoints = colors.resolveBreakpoints;
exports.resolveVerticalBreakpoints = colors.resolveVerticalBreakpoints;
exports.valueHandlers = colors.valueHandlers;
exports.xyzMap = colors.xyzMap;
exports.variantMatcher = variants.variantMatcher;
exports.variantParentMatcher = variants.variantParentMatcher;
