// unocss-preset-weapp/dist/variants.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const _default = require('./shared/unocss-preset-weapp.4c6357e3.cjs');
require('./shared/unocss-preset-weapp.80bd545b.cjs');
require('@unocss/core');
require('unplugin-transform-class/utils');
require('./shared/unocss-preset-weapp.60967b4e.cjs');



exports.partClasses = _default.partClasses;
exports.variantBreakpoints = _default.variantBreakpoints;
exports.variantColorsMediaOrClass = _default.variantColorsMediaOrClass;
exports.variantCombinators = _default.variantCombinators;
exports.variantCssLayer = _default.variantCssLayer;
exports.variantCustomMedia = _default.variantCustomMedia;
exports.variantImportant = _default.variantImportant;
exports.variantInternalLayer = _default.variantInternalLayer;
exports.variantLanguageDirections = _default.variantLanguageDirections;
exports.variantNegative = _default.variantNegative;
exports.variantPrint = _default.variantPrint;
exports.variantPseudoClassFunctions = _default.variantPseudoClassFunctions;
exports.variantPseudoClassesAndElements = _default.variantPseudoClassesAndElements;
exports.variantScope = _default.variantScope;
exports.variantSelector = _default.variantSelector;
exports.variantTaggedPseudoClasses = _default.variantTaggedPseudoClasses;
exports.variantVariables = _default.variantVariables;
exports.variants = _default.variants;
